certbot/ocsp.py

```diff
@@ -65,33 +88,128 @@ def ocsp_revoked(self, cert_path, chain_path):
+    if response.status_code != 200:
+        logger.info("OCSP check failed for %s (are we offline?)", cert_path)
+        return False
+    response_ocsp = ocsp.load_der_ocsp_response(response.content)
+
+    # Check OCSP response validity
+    if response_ocsp.response_status != ocsp.OCSPResponseStatus.SUCCESSFUL:
+        logger.error("Invalid OCSP response status for %s: %s",
+                     cert_path, response_ocsp.response_status)
+        return False
+
+    # Check OCSP signature
+    try:
+        _check_ocsp_response(response_ocsp, request, issuer)
+    except UnsupportedAlgorithm as e:
+        logger.error(str(e))
+        return False
+    except InvalidSignature:
+        logger.error('Invalid signature on OCSP response for %s', cert_path)
+        return False
+    except AssertionError as error:
+        logger.error('Invalid OCSP response for %s: %s.', cert_path, str(error))
+        return False
+
+    # Check OCSP certificate status
+    logger.debug("OCSP certificate status for %s is: %s",
+                 cert_path, response_ocsp.certificate_status)
+    return response_ocsp.certificate_status == ocsp.OCSPCertStatus.REVOKED
+
+
+def _check_ocsp_response(response_ocsp, request_ocsp, issuer_cert):
+    """Verify that the OCSP is valid for serveral criterias"""
+    # Assert OCSP response corresponds to the certificate we are talking about
+    if response_ocsp.serial_number != request_ocsp.serial_number:
+        raise AssertionError('the certificate in response does not correspond '
+                             'to the certificate in request')
+
+    # Assert signature is valid
+    _check_ocsp_response_signature(response_ocsp, issuer_cert)
+
+    # Assert issuer in response is the expected one
+    if response_ocsp.issuer_name_hash != request_ocsp.issuer_name_hash:
+        raise AssertionError('the issuer does not correspond to issuer of the certificate.')
+
+    # Assert nextUpdate is in the future, and that thisUpdate is not too old
+    if response_ocsp.next_update:
+        if response_ocsp.next_update < datetime.now():
+            raise AssertionError('next update is in the past.')
+        interval = response_ocsp.next_update - response_ocsp.this_update
+        if datetime.now() - response_ocsp.this_update > interval:
+            raise AssertionError('this update is too old.')
+
+
+def _check_ocsp_response_signature(response_ocsp, issuer_cert):
+    """Verify an OCSP response signature against certificate issuer"""
+    try:
+        # TODO: (adferrand 2019-11-01) Following line can be improved using a direct call to
+        #  response_ocsp.signature_hash_algorithm once cryptography 2.5 is released
+        #  (watch out for retro-compatibility with 2.4 though).
+        #  See https://github.com/pyca/cryptography/issues/4680
+        chosen_hash = x509._SIG_OIDS_TO_HASH[response_ocsp.signature_algorithm_oid]  # pylint: disable=protected-access
```

Comment:
After thinking about this a little more, I'm hesitant for us to ship code relying on private functionality like this. I just think it can break too easily and cause problems.

How about we disable this code from being used for production, but have a way to enable using `cryptography` during tests? Ideally these tests should check specifically for the version we're pinning to avoid potential problems for packagers running our test suite. After `cryptography` 2.5 is released, we can enable this feature in production using `signature_hash_algorithm`.

Reply:
I do not see how to easily make a specific behavior on this code for tests only. And adding code for tests in the software itself is bad in my opinion.

Let's consider doing this logic: if there is signature_hash_algorithm method, use it, otherwise use _SIG_OIDS_TO_HASH. And as soon as cryptography 2.5 lands, we escalate the minimal version for cryptography to 2.5 on Certbot, and remove the execution on _SIG_OIDS_TO_HASH to use signature_hash_algorithm only. Then we will have a unique version of Certbot, the next one, that in case a unique version of cryptography is installed in the OS, 2.4, will rely on the internal field, that, at the time of 2.4, exists and behaves correctly.

For any other situation, either openssl will be used, or cryptography will be used through signature_hash_algorithm.

I think it is a pretty safe situation.